### ISSUE TITLE
GCop 441

### ISSUE THREAD

**Issue by marjanjavid** [Geeksltd/GCop#231]
this rule emphasize on using `is` operator instead of `as` operator **when we want to determine an object's type but do not have to actually cast it**. So it seems in the sample below this rule should not fire but is fired in this case:
```csharp
if (foo as string != null)
{
      var bar = foo as string;
}
```
Do we need to make difference for this or not?

Also pay attention that in C# 7.0 and later, we can use the `is` operator with pattern matching to check the type conversion and cast the expression to a variable of that type in one step(this is useful for our warning message).This can be a new rule too

also please read 
https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1800-do-not-cast-unnecessarily?view=vs-2017

**Comment by Foroughi** [comment]
@marjanjavid To me, the rule SHOULD kick in, in that sample as its better to write it like this : 

```
if (foo is string)
{
      var bar = foo as string;
}
```
Aint that true? or I'm missing something here?

**Comment by marjanjavid** [comment]
@Foroughi Yes actually you are right but in a link I mentioned above (which is a Microsoft doc) we have a paragraph that says:

**If the C# is operator is used to test whether the cast will succeed before the actual cast is performed, consider testing the result of the as operator instead. This provides the same functionality without the implicit cast operation that is performed by the is operator. Or, in C# 7.0 and later, use the is operator with pattern matching to check the type conversion and cast the expression to a variable of that type in one step.**

also it has a sample that is like what u write above:
```CSharp
// The 'is' statement performs a cast operation.
 if(obj is Control) 
       {
             // The 'as' statement performs a duplicate cast operation.
             Control aControl = obj as Control;
             // Use aControl.
        }
```
As what this article says, we should change this rule like below example because we are casting after checking (it says that we should not use `is`, to prevent double cast):
```CSharp
             Control aControl = obj as Control;
              if (aControl != null)
              {
                  // Use aControl.
              }
```
or in C#7 we can write it like below
```CSharp
            if (obj is Control aControl)
              {
                  // Use aControl.
                  Console.WriteLine(aControl.Name);
              }
```
As a result we should check in GCop, if we have a cast in the body of If-Condition for the object written in `if ` condition, then we should not use `is`.

**Comment by marjanjavid** [comment]
It works fine on GCop2.6.0
